src/unix/stream.c

```diff
@@ -1050,16 +1050,21 @@ static void uv__read(uv_stream_t* stream) {
 
   is_ipc = stream->type == UV_NAMED_PIPE && ((uv_pipe_t*) stream)->ipc;
 
-  /* XXX: Maybe instead of having UV_HANDLE_READING we just test if
-   * tcp->read_cb is NULL or not?
-   */
-  while (stream->read_cb
-      && (stream->flags & UV_HANDLE_READING)
-      && (count-- > 0)) {
-    assert(stream->alloc_cb != NULL);
+  while (count-- > 0) {
+    if (!(stream->flags & UV_HANDLE_READING))
+      return;
 
+    assert(stream->alloc_cb != NULL);
     buf = uv_buf_init(NULL, 0);
     stream->alloc_cb((uv_handle_t*)stream, 64 * 1024, &buf);
```

Comment:
actually... there's still some weird behavior in the pathological case where a user calls `read_stop(); read_start()` with different callbacks inside the alloc_cb, which is that you might not detect that read_stop was canceled, and you might end up passing an allocation from one alloc_cb into a totally different read_cb.

The only thing I can think of to detect that is to cache both the `stream->alloc_cb` and `stream->read_cb` before calling alloc_cb, and detect if either have changed or if read_stop was called.

It's a weird enough situation that you could probably outlaw it instead of solving it.

Reply:
"`alloc_cb` is only for allocating; calling `uv_read_stop`, `uv_read_start`, or `uv_close` within the `alloc_cb` is not supported and will result in undefined behavior, which may include memory leaks.  Instead, make those calls in the subsequent `read_cb` which follows every `alloc_cb`"

This feels like a reasonable restriction to me, and the fact that this issue has not come up before now tells me that nearly everybody is already following this rule anyway.